// @patternfly/react-topology/dist/js/behavior/useAreaSelection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withAreaSelection = exports.useAreaSelection = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const d3 = tslib_1.__importStar(require("d3"));
const mobx_react_1 = require("mobx-react");
const mobx_1 = require("mobx");
const ElementContext_1 = tslib_1.__importDefault(require("../utils/ElementContext"));
const useCallbackRef_1 = tslib_1.__importDefault(require("../utils/useCallbackRef"));
const types_1 = require("../types");
const Point_1 = tslib_1.__importDefault(require("../geom/Point"));
// Used to send events prevented by d3.zoom to the document allowing modals, dropdowns, etc, to close
const propagateAreaSelectionMouseEvent = (e) => {
    document.dispatchEvent(new MouseEvent(e.type, e));
};
const useAreaSelection = (modifiers = ['ctrlKey']) => {
    const element = (0, react_1.useContext)(ElementContext_1.default);
    const [draggingState, setDraggingState] = (0, react_1.useState)({});
    if (!(0, types_1.isGraph)(element)) {
        throw new Error('useAreaSelection must be used within the scope of a Graph');
    }
    const elementRef = (0, react_1.useRef)(element);
    elementRef.current = element;
    const areaSelectionRef = (0, useCallbackRef_1.default)((node) => {
        if (node) {
            // TODO fix any type
            const $svg = d3.select(node.ownerSVGElement);
            if (node && node.ownerSVGElement) {
                node.ownerSVGElement.addEventListener('mousedown', propagateAreaSelectionMouseEvent);
                node.ownerSVGElement.addEventListener('click', propagateAreaSelectionMouseEvent);
            }
            const drag = d3
                .drag()
                .on('start', (0, mobx_1.action)((event) => {
                const { offsetX, offsetY } = event.sourceEvent instanceof MouseEvent ? event.sourceEvent : { offsetX: 0, offsetY: 0 };
                const { width: maxX, height: maxY } = elementRef.current.getDimensions();
                const startPoint = new Point_1.default(Math.min(Math.max(offsetX, 0), maxX), Math.min(Math.max(offsetY, 0), maxY));
                const modifier = modifiers.find((m) => event.sourceEvent[m]);
                setDraggingState({
                    modifier,
                    isAreaSelectDragging: true,
                    areaSelectDragStart: startPoint,
                    areaSelectDragEnd: startPoint
                });
                elementRef.current
                    .getController()
                    .fireEvent(types_1.GRAPH_AREA_DRAGGING_EVENT, { graph: elementRef.current, isDragging: true });
            }))
                .on('drag', (0, mobx_1.action)((event) => {
                const { offsetX, offsetY } = event.sourceEvent instanceof MouseEvent ? event.sourceEvent : { offsetX: 0, offsetY: 0 };
                const { width: maxX, height: maxY } = elementRef.current.getDimensions();
                setDraggingState((prev) => (Object.assign(Object.assign({}, prev), { areaSelectDragEnd: new Point_1.default(Math.min(Math.max(offsetX, 0), maxX), Math.min(Math.max(offsetY, 0), maxY)) })));
            }))
                .on('end', (0, mobx_1.action)(() => {
                setDraggingState((prev) => {
                    elementRef.current.getController().fireEvent(types_1.GRAPH_AREA_SELECTED_EVENT, {
                        graph: elementRef.current,
                        modifier: prev.modifier,
                        startPoint: prev.areaSelectDragStart,
                        endPoint: prev.areaSelectDragEnd
                    });
                    return { isAreaSelectDragging: false };
                });
                elementRef.current
                    .getController()
                    .fireEvent(types_1.GRAPH_AREA_DRAGGING_EVENT, { graph: elementRef.current, isDragging: false });
            }))
                .filter((event) => modifiers.find((m) => event[m]) && !event.button);
            drag($svg);
        }
        return () => {
            if (node) {
                // remove all drag listeners
                d3.select(node.ownerSVGElement).on('.drag', null);
                if (node.ownerSVGElement) {
                    node.ownerSVGElement.removeEventListener('mousedown', propagateAreaSelectionMouseEvent);
                    node.ownerSVGElement.removeEventListener('click', propagateAreaSelectionMouseEvent);
                }
            }
        };
    });
    return Object.assign({ areaSelectionRef }, draggingState);
};
exports.useAreaSelection = useAreaSelection;
const withAreaSelection = (modifier = ['ctrlKey']) => (WrappedComponent) => {
    const Component = (props) => {
        const areaSelectionProps = (0, exports.useAreaSelection)(modifier);
        return (0, jsx_runtime_1.jsx)(WrappedComponent, Object.assign({}, props, areaSelectionProps));
    };
    Component.displayName = `withAreaSelection(${WrappedComponent.displayName || WrappedComponent.name})`;
    return (0, mobx_react_1.observer)(Component);
};
exports.withAreaSelection = withAreaSelection;
