// @openshift-console/dynamic-plugin-sdk/lib/extensions/console-types.js
var InventoryStatusGroup;
(function (InventoryStatusGroup) {
    InventoryStatusGroup["WARN"] = "WARN";
    InventoryStatusGroup["ERROR"] = "ERROR";
    InventoryStatusGroup["PROGRESS"] = "PROGRESS";
    InventoryStatusGroup["NOT_MAPPED"] = "NOT_MAPPED";
    InventoryStatusGroup["UNKNOWN"] = "UNKNOWN";
})(InventoryStatusGroup || (InventoryStatusGroup = {}));
export var HealthState;
(function (HealthState) {
    HealthState["OK"] = "OK";
    HealthState["ERROR"] = "ERROR";
    HealthState["WARNING"] = "WARNING";
    HealthState["LOADING"] = "LOADING";
    HealthState["UNKNOWN"] = "UNKNOWN";
    HealthState["UPDATING"] = "UPDATING";
    HealthState["PROGRESS"] = "PROGRESS";
    HealthState["UPGRADABLE"] = "UPGRADABLE";
    HealthState["NOT_AVAILABLE"] = "NOT_AVAILABLE";
})(HealthState || (HealthState = {}));
export var GridPosition;
(function (GridPosition) {
    GridPosition["MAIN"] = "MAIN";
    GridPosition["LEFT"] = "LEFT";
    GridPosition["RIGHT"] = "RIGHT";
})(GridPosition || (GridPosition = {}));
export var LIMIT_STATE;
(function (LIMIT_STATE) {
    LIMIT_STATE["ERROR"] = "ERROR";
    LIMIT_STATE["WARN"] = "WARN";
    LIMIT_STATE["OK"] = "OK";
})(LIMIT_STATE || (LIMIT_STATE = {}));
export var K8sResourceConditionStatus;
(function (K8sResourceConditionStatus) {
    K8sResourceConditionStatus["True"] = "True";
    K8sResourceConditionStatus["False"] = "False";
    K8sResourceConditionStatus["Unknown"] = "Unknown";
})(K8sResourceConditionStatus || (K8sResourceConditionStatus = {}));
export var AllPodStatus;
(function (AllPodStatus) {
    AllPodStatus["Running"] = "Running";
    AllPodStatus["NotReady"] = "Not Ready";
    AllPodStatus["Warning"] = "Warning";
    AllPodStatus["Empty"] = "Empty";
    AllPodStatus["Failed"] = "Failed";
    AllPodStatus["Pending"] = "Pending";
    AllPodStatus["Succeeded"] = "Succeeded";
    AllPodStatus["Terminating"] = "Terminating";
    AllPodStatus["Unknown"] = "Unknown";
    AllPodStatus["ScaledTo0"] = "Scaled to 0";
    AllPodStatus["Idle"] = "Idle";
    AllPodStatus["AutoScaledTo0"] = "Autoscaled to 0";
    AllPodStatus["ScalingUp"] = "Scaling Up";
    AllPodStatus["CrashLoopBackOff"] = "CrashLoopBackOff";
})(AllPodStatus || (AllPodStatus = {}));
