// @patternfly/react-topology/dist/js/components/popper/Popper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const popper_js_1 = tslib_1.__importDefault(require("popper.js"));
const useCombineRefs_1 = tslib_1.__importDefault(require("../../utils/useCombineRefs"));
const Portal_1 = tslib_1.__importDefault(require("./Portal"));
class VirtualReference {
    constructor({ height = 0, width = 0, x, y }) {
        this.rect = {
            bottom: y + height,
            height,
            left: x,
            right: x + width,
            top: y,
            width,
            x,
            y,
            toJSON: () => { }
        };
    }
    getBoundingClientRect() {
        return this.rect;
    }
    get clientWidth() {
        return this.rect.width || 0;
    }
    get clientHeight() {
        return this.rect.height || 0;
    }
}
const getReference = (reference) => 'getBoundingClientRect' in reference ? reference : new VirtualReference(reference);
const DEFAULT_POPPER_OPTIONS = {};
const Popper = ({ children, container, className, open, placement = 'bottom-start', reference, popperOptions = DEFAULT_POPPER_OPTIONS, closeOnEsc, closeOnOutsideClick, onRequestClose, popperRef: popperRefIn, zIndex = 9999, returnFocus }) => {
    const controlled = typeof open === 'boolean';
    const openProp = controlled ? open || false : true;
    const nodeRef = (0, react_1.useRef)(null);
    const popperRef = (0, react_1.useRef)(null);
    const popperRefs = (0, useCombineRefs_1.default)(popperRef, popperRefIn);
    const [isOpen, setOpenState] = (0, react_1.useState)(openProp);
    const focusRef = (0, react_1.useRef)(null);
    const onRequestCloseRef = (0, react_1.useRef)(onRequestClose);
    onRequestCloseRef.current = onRequestClose;
    const setOpen = (0, react_1.useCallback)((newOpen) => {
        if (returnFocus && newOpen !== isOpen) {
            if (newOpen) {
                if (document.activeElement) {
                    focusRef.current = document.activeElement;
                }
            }
            else if (focusRef.current instanceof HTMLElement && focusRef.current.ownerDocument) {
                focusRef.current.focus();
            }
        }
        setOpenState(newOpen);
    }, [returnFocus, isOpen]);
    (0, react_1.useEffect)(() => {
        setOpen(openProp);
    }, [openProp, setOpen]);
    const onKeyDown = (0, react_1.useCallback)((e) => {
        if (e.key === 'Escape') {
            controlled ? onRequestCloseRef.current && onRequestCloseRef.current() : setOpen(false);
        }
    }, [controlled, setOpen]);
    const onClickOutside = (0, react_1.useCallback)((e) => {
        if (!nodeRef.current || (e.target instanceof Node && !nodeRef.current.contains(e.target))) {
            controlled ? onRequestCloseRef.current && onRequestCloseRef.current(e) : setOpen(false);
        }
    }, [controlled, setOpen]);
    const destroy = (0, react_1.useCallback)(() => {
        if (popperRef.current) {
            popperRef.current.destroy();
            popperRefs(null);
            document.removeEventListener('keydown', onKeyDown, true);
            document.removeEventListener('mousedown', onClickOutside, true);
            document.removeEventListener('touchstart', onClickOutside, true);
        }
    }, [onClickOutside, onKeyDown, popperRefs]);
    const initialize = (0, react_1.useCallback)(() => {
        if (!nodeRef.current || !reference || !isOpen) {
            return;
        }
        destroy();
        popperRefs(new popper_js_1.default(getReference(typeof reference === 'function' ? reference() : reference), nodeRef.current, Object.assign(Object.assign({ placement }, popperOptions), { modifiers: Object.assign({ preventOverflow: {
                    boundariesElement: 'viewport'
                } }, popperOptions.modifiers) })));
        // init document listenerrs
        if (closeOnEsc) {
            document.addEventListener('keydown', onKeyDown, true);
        }
        if (closeOnOutsideClick) {
            document.addEventListener('mousedown', onClickOutside, true);
            document.addEventListener('touchstart', onClickOutside, true);
        }
    }, [
        popperRefs,
        reference,
        isOpen,
        destroy,
        placement,
        popperOptions,
        closeOnEsc,
        closeOnOutsideClick,
        onKeyDown,
        onClickOutside
    ]);
    const nodeRefCallback = (0, react_1.useCallback)((node) => {
        nodeRef.current = node;
        initialize();
    }, [initialize]);
    (0, react_1.useEffect)(() => {
        initialize();
    }, [initialize]);
    (0, react_1.useEffect)(() => () => {
        destroy();
    }, [destroy]);
    (0, react_1.useEffect)(() => {
        if (!isOpen) {
            destroy();
        }
    }, [destroy, isOpen]);
    return isOpen ? ((0, jsx_runtime_1.jsx)(Portal_1.default, Object.assign({ container: container }, { children: (0, jsx_runtime_1.jsx)("div", Object.assign({ ref: nodeRefCallback, className: className, style: { zIndex } }, { children: children })) }))) : null;
};
exports.default = Popper;
