// @openshift-console/dynamic-plugin-sdk/lib/app/core/actions/core.js
import { action } from 'typesafe-actions';
export var ActionType;
(function (ActionType) {
    ActionType["SetUser"] = "setUser";
    ActionType["SetUserResource"] = "setUserResource";
    ActionType["BeginImpersonate"] = "beginImpersonate";
    ActionType["EndImpersonate"] = "endImpersonate";
    ActionType["SetActiveCluster"] = "setActiveCluster";
    ActionType["SetAdmissionWebhookWarning"] = "setAdmissionWebhookWarning";
    ActionType["RemoveAdmissionWebhookWarning"] = "removeAdmissionWebhookWarning";
})(ActionType || (ActionType = {}));
export const setUser = (userInfo) => action(ActionType.SetUser, { userInfo });
export const setUserResource = (userResource) => action(ActionType.SetUserResource, { userResource });
export const beginImpersonate = (kind, name, subprotocols, groups) => action(ActionType.BeginImpersonate, { kind, name, subprotocols, groups });
export const endImpersonate = () => action(ActionType.EndImpersonate);
export const setAdmissionWebhookWarning = (id, warning) => action(ActionType.SetAdmissionWebhookWarning, { id, warning });
export const removeAdmissionWebhookWarning = (id) => action(ActionType.RemoveAdmissionWebhookWarning, { id });
const coreActions = {
    setUser,
    setUserResource,
    beginImpersonate,
    endImpersonate,
    setAdmissionWebhookWarning,
    removeAdmissionWebhookWarning,
};
